Improve TZ selector markup and UX

foxbunny opened the pull request:
- Modify the TZ selector markup to be more keyboard accessible
- Modify the markup to convey information about states
- Modify the TZ select list behavior to automatically activate the "Custom" radio when selection is made
- Modify the TZ selector to not disable the select list when options other than "Custom" are used

This patch also adds some initial code for a design system.

Fixes #5908

foxbunny commented:
@ThiefMaster Heads-up: this patch includes an experimental design system that I propose for the use on public pages. The code is contained inside the `styles/design_system` folder (there is a README) file in there that explains what the intent is and how it should work. The DS already addresses some of the color contrast issues, but it generally tries to approximately match the existing UI and integrate with it. Please check it out and let me know your thoughts.

ThiefMaster commented:
Not sure I'll find the time to look into it today and I'm off tomorrow, but I'll have a closer look next week, but maybe @tomasr8 can already provide some input on it in the meantime

foxbunny commented:
> Not sure I'll find the time to look into it today and I'm off tomorrow, but I'll have a closer look next week, but maybe @tomasr8 can already provide some input on it in the meantime

No rush

tomasr8 commented:
Haven't gone through the DS yet, but just noticed a small error when navigating to the management area of an event (probably because the tz selector is disabled there):

```js
Uncaught TypeError: $tzCustomField is null
    js tz_selector.js:18
    js tz_selector.js:9
```

foxbunny commented:
> noticed a small error when navigating to the management area

Should be fixed now

foxbunny commented:
Not sure what this CI failure is about. Doesn't look like it's related to the changes I've made...

tomasr8 commented:
> Not sure what this CI failure is about. Doesn't look like it's related to the changes I've made...

Maybe some GH hiccup? I tried rerunning the job and it failed again..

ThiefMaster commented:
Yeah, the broken python interpreter happens from time to time on github... annoying but harmless

ThiefMaster commented:
CI bug is fixed in #5920, rebase to the latest master and it should be fine again

ThiefMaster commented on indico/web/client/styles/design_system/README.txt:
```diff
@@ -0,0 +1,114 @@
+Indico Design System
+====================
+
+
+This file is part of Indico.
+Copyright (C) 2002 - 2023 CERN
+
+Indico is free software; you can redistribute it and/or modify it
+under the terms of the MIT License; see the LICENSE file for more
+details.
+
```
not needed in a random README (also because this would not be caught by our tool to update headers)

ThiefMaster commented on indico/web/client/styles/design_system/_index.scss:
```diff
@@ -0,0 +1,15 @@
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+@use 'variables';
+@use 'button';
+@use 'button_group';
+@use 'form_controls';
+@use 'layout';
+@use 'popup';
+@use 'visibility';
+@use 'utils';
```
I'd be clear using `./whatever` that those are relative imports and not some top-level thing (like `'base'` would be for example)

tomasr8 commented on indico/web/client/styles/design_system/README.txt:
```diff
@@ -0,0 +1,114 @@
+
+The intended usage is as follows:
+
+    @import "design_system";
```
Any reason for `@import` instead of `@use`?

foxbunny commented:
The only caveat with placeholders is that each media query is essentially a separate context. That may result in counter-intuitive behavior. For example:

```
@media (width < 70rem) {
  foo {
    @extend %flex-column;
  }
}
```

This doesn't work because the placeholder itself is *outside* the media target. This is basically a trade-off of using this style of code reuse. For this, we would do something like this:

```
mixin flex-column {
  display: flex;
  flex-direction: column;
}

%flex-column {
  @include flex-column;
}
```

So we'd have both a mixin and a placeholder. The media query would result in some code duplication, but it shouldn't be too bad overall.

tomasr8 commented:
> The only caveat with placeholders is that each media query is essentially a separate context. That may result in counter-intuitive behavior.

I think this is worth mentioning in the readme

foxbunny commented:
Good call. I'll fix it first thing tomorrow.

ThiefMaster commented on indico/web/client/styles/design_system/README.txt:
```diff
@@ -0,0 +1,157 @@
+Indico Design System
```
I think this file would deserve being markdown as well :)

ThiefMaster replied:
And just thinking, would it make sense to move the design system in a separate PR? I think both things are pretty much ready to merge, but it might be nicer to have them in separate PRs (also because IMHO the design system deserves a changelog entry in 'internal changes')

foxbunny replied:
> And just thinking, would it make sense to move the design system in a separate PR?

To me this type of decision is usually based on the need to revert them separately. I've no strong opinion either way.

ThiefMaster replied:
Same here (no strong opinion). If it's already a separate commit I'd split it, if not I wouldn't bother to spend/waste time with it ;)

ThiefMaster commented:
![image](https://github.com/indico/indico/assets/179599/deb3d609-f074-4493-94ff-0d436cbc0b83)

In the old version, the timezone select dropdown is disabled when one of the first two options are checked. This is no longer the case - intentionally?

You also no longer see what your default profile timezone is - maybe the dropdown should switch to your timezone when selecting that option?

One case of unexpected behavior is if you select a custom timezone, enable 'remember these settings', and then click on "event/category" or "profile" again and save. Then it still updates your default timezone in the database (tip: run `python bin/utils/db_log.py -S` to see all non-SELECT queries being executed)

foxbunny commented:
> In the old version, the timezone select dropdown is disabled when one of the first two options are checked. This is no longer the case - intentionally?

Yes. Totally intentional.

> You also no longer see what your default profile timezone is - maybe the dropdown should switch to your timezone when selecting that option?

Ok, I'll figure something out. Probably a message below the buttons or something.

> One case of unexpected behavior is if you select a custom timezone, enable 'remember these settings', and then click on "event/category" or "profile" again and save. Then it still updates your default timezone in the database (tip: run python bin/utils/db_log.py -S to see all non-SELECT queries being executed)

If I understand this correctly, "remember these settings" basically means "use custom setting as your profile timezone"?

foxbunny commented:
I think I see what the issue is. Since you're checking `tz and tz in common_timezones_set`, and the new implementation doesn't disable this field, it always sets the user time zone regardless of what mode you're in.

I'll fix this on Monday.

foxbunny commented:
Updated PR.

There is now an indication of your profile timezone:

<img width="264" alt="image" src="https://github.com/indico/indico/assets/65413/0a5d1b59-1666-401e-bc4c-011ef90db4df">

The profile timezone is also highlighted in the custom timezone list, with an additional "(your profile timezone)" label:

<img width="240" alt="image" src="https://github.com/indico/indico/assets/65413/62c2d1c4-1abd-4e9c-8c18-7c51236a4b61">


This item will scroll into view when "Profile" option is selected.

When the "Custom" option is selected, any selected item in the select list will be scrolled into view.

I've fixed the behavior of the "Remember these settings" by making a change in the controller as follows:


```
        elif mode == 'custom' and tz in common_timezones_set:
            session.timezone = tz
            if update_user and session.user:  # <--- MOVED HERE
                session.user.settings.set('timezone', tz)

        if update_user and session.user:
            session.user.settings.set('force_timezone', mode != 'local')
```

This now makes the backend behavior match the UI. Please let me know if this fix is appropriate.

foxbunny commented:
Also, converted README to markdown.

foxbunny commented:
Not quite sure why these tests are failing. I tried running locally, but I get a different result. What command should I use to run the python tests?

ThiefMaster commented:
It's also broken when you simply access indico without being logged in (`session.user` is None in that case)

tomasr8 commented:
`pytest --no-cov indico/web/util_test.py` to reproduce ;)

ThiefMaster commented on indico/web/templates/_session_bar.html:
```diff
@@ -48,64 +48,76 @@
+                                <option value="{{ tz }}" {{ 'selected' if tz == session.tzinfo.zone }} {{ 'data-profile' if is_profile_timezone }}>
+                                    {{ tz }}
+                                    {% if is_profile_timezone -%}
+                                        {# Translators: example usage: "Europe/Geneva (your profile timezone)" #}
```
Nobody will read that ;x

I suggest doing this instead (untested):

```jinja
                                    {% if is_profile_timezone -%}
                                        {% trans %}{{ tz }} (your profile timezone){% endtrans %}
                                    {%- else -%}
                                        {{- tz -}}
                                    {%- endif %}                                    
```

That way the placeholder is in the translatable string so it'll be much clearer how it's used

foxbunny replied:
Ah, great idea! Thanks

foxbunny commented:
> `pytest --no-cov indico/web/util_test.py` to reproduce ;)

I get a bunch of these:

```
SKIPPED [1] indico/web/util_test.py:598: Could not retrieve PostgreSQL version: [Errno 2] No such file or directory: 'initdb'
```

foxbunny commented:
Any idea why this test in particular would fail for me?

```
_________________________ test_proxy_to_reservation_if_last_valid_occurrence[False-True-True-True] _________________________

db = <IndicoSQLAlchemy postgresql://indico:indico@localhost:5432/indico_test>
mocker = <pytest_mock.plugin.MockerFixture object at 0x7fed88f1ea90>
create_reservation = <function create_reservation.<locals>._create_reservation at 0x7fed88f17b80>
dummy_user = <User(1337, 1337@example.test): "Guinea Pig">, not_repeating = False, only_one_valid = True, propagate = True
proxied = True, freeze_time = <function freeze_time.<locals>._freeze_time at 0x7fed88f17e50>

    @pytest.mark.parametrize(
        ('not_repeating', 'only_one_valid', 'propagate', 'proxied'),
        bool_matrix('..0', expect=False) +
        bool_matrix('..1', expect='any_dynamic')
    )
    @pytest.mark.usefixtures('smtp')
    def test_proxy_to_reservation_if_last_valid_occurrence(db, mocker, create_reservation, dummy_user,
                                                           not_repeating, only_one_valid, propagate, proxied, freeze_time):
        resv = create_reservation(start_dt=datetime.combine(date.today(), time(8)),
                                  end_dt=datetime.combine(date.today() + timedelta(days=1), time(17)),
                                  repeat_frequency=RepeatFrequency.NEVER if not_repeating else RepeatFrequency.DAY)
        freeze_time(datetime.combine(date.today(), time(8)))
        if only_one_valid:
            for occ in resv.occurrences[1:]:
                occ.state = ReservationOccurrenceState.cancelled
            db.session.flush()

        occ = resv.occurrences.first()
        mocker.patch.object(resv, 'cancel')
        occ.cancel(user=dummy_user, propagate=propagate)
>       assert resv.cancel.called == proxied
E       assert False == True
E        +  where False = <MagicMock name='cancel' id='140658181178128'>.called
E        +    where <MagicMock name='cancel' id='140658181178128'> = <Reservation(249, room_id=1133, 2023-09-25 08:00:00, 2023-09-26 17:00:00, accepted): "Testing">.cancel

indico/modules/rb/models/util_test.py:40: AssertionError
```

foxbunny commented:
Rebased on master

ThiefMaster commented:
no idea about the failing test. a wild guess would be that the failure is depending on the time of the day when it runs

foxbunny commented:
I currently get no failures after rebasing on master.

ThiefMaster commented on indico/web/client/js/widgets/tz_selector.js:
```diff
@@ -0,0 +1,35 @@
+// LICENSE file for more details.
+
+// eslint-disable-next-line import/unambiguous
+window.addEventListener('DOMContentLoaded', function() {
```
```suggestion
window.addEventListener('DOMContentLoaded', () => {
```

foxbunny replied:
Btw, I think arrow callbacks can be enforced via ESLint.

ThiefMaster replied:
I [thought so too](https://github.com/indico/eslint-config-indico/blob/909231c41a7a347b85925f720f2715a967b91696/configs/main.yaml#L162C3-L162C30)... `prefer-arrow-callback: warn`

foxbunny commented on indico/web/client/js/widgets/tz_selector.js:
```diff
@@ -0,0 +1,35 @@
+    customRadio.checked = true;
+  }
+
+  tzCustomField?.addEventListener('click', setCustomMode);
+  tzCustomField?.addEventListener('change', setCustomMode);
```
Yeah, it's completely unnecessary. Just forgot to remove them.

foxbunny commented on indico/web/client/styles/widgets/_tz_selector.scss:
```diff
@@ -0,0 +1,80 @@
+    flex: 1;
+  }
+
+  [type='radio'] {
```
1. I think it's better to underspecify a selector whenever possible, as that keeps them more flexible. For instance, I use `[name="foo"]`, `[role="tab"]` and similar selectors where applicable.
2. If efficiency of a selector ever becomes an issue, it probably means there's something deeply wrong with the UI elsewhere :)

foxbunny commented on indico/web/client/styles/widgets/_tz_selector.scss:
```diff
@@ -0,0 +1,80 @@
+
+@use 'design_system';
+
+// NB: The following rule is a temporary hack to work around QTip2 styling. It should be removed after the popup is converted into a pull-down menu.
```
Is "XXX" your convention for "todo" items?

ThiefMaster replied:
it's more for general notes. TODO is for actual todo items

foxbunny commented:
Addressed comments

ThiefMaster commented:
@foxbunny I think you missed some (maybe collapsed by github?) :/

foxbunny commented:
Ok, will check tomorrow

foxbunny commented:
> I think you missed some

I went over the last batch of comments, and couldn't find what you're referring to. Can you please point me to the comments?

ThiefMaster commented on indico/web/client/styles/design_system/README.md:
```diff
@@ -0,0 +1,163 @@
+This is decidedly different from the common way of using CSS in which the
+appearance of the element is defined using classes. Instead of adding
+classes to elements in order to specify their appearance, we instead use
+placeholder selectors and @extend to specify the appearance for elements
```
@foxbunny ^

ThiefMaster commented on indico/web/client/styles/design_system/README.md:
````diff
@@ -0,0 +1,163 @@
+```
+
+If SCSS were to compile the code the way we instructed, it would break the
+media query. Within @rules, we still need to use mixins. In most cases, this
````
@foxbunny ^

ThiefMaster commented on indico/web/client/styles/design_system/_utils.scss:
```diff
@@ -0,0 +1,16 @@
+}
+
+%default-focus-ring {
+  outline: medium solid Highlight;
```
@foxbunny ^

foxbunny replied:
It's from here:

https://css-tricks.com/copy-the-browsers-native-focus-styles/

<img width="399" alt="image" src="https://github.com/indico/indico/assets/65413/53075c08-dbf8-403a-9119-621d62eb20f6">

ThiefMaster commented on indico/web/client/styles/partials/_icons.scss:
```diff
@@ -151,7 +152,8 @@ i[class^='icon-']::before {
 .icon-checkbox-unchecked::before {
   content: '\e433';
 }
-.icon-checkmark::before {
+.icon-checkmark::before,
+%icon-checkmark {
```
@foxbunny ^

foxbunny replied:
Ok, let me see what I can do.

foxbunny replied:
Btw, I was intending to duplicate all of those as placeholders. I can probably come up with a way to automate that so that it's not (as) painful to update.

foxbunny replied:
Ok, there we go.

https://codepen.io/hayavuk/pen/KKbGzbW

I'll include this in the instructions.

foxbunny replied:
@ThiefMaster Please try it before I push the changes, let me know if it works for you.

foxbunny replied:
What it does is:

- Changes all selectors to use dashes instead of underscores
- Lower-cases all selectors
- Changes double to single quotes
- Adds the placeholders

foxbunny replied:
I'm noticing that the icomoon JSON you have in the repo contains some icons that aren't in the _icons.scss. Just an FYI.

ThiefMaster replied:
The extra ones are just alias names that are added by default. The unfortunately need to be removed manually. Alternatively, maybe only the first class name for any given `content` value could be kept since that's already the one we're using anyway.

Your script to convert it works fine. I'd suggest to add it in `bin/maintenance/` though, to avoid relying on some external codepen.

It could actually work directly on the _icons.scss and modify the parts between `/* Icomoon's icon definitions */` and `/* End of icomoon's icon definitions */` (so you'd just paste the icomoon css there and then run that script to fix things)

ThiefMaster replied:
Actually, even better than pasting into that area would be the script taking the icomoon css file as input (or having the user paste it into the script's stdin). In either case I would change the begin/end markers to something where it's clearer that it's meant for a machine and not a human.

`/* BEGIN GENERATED ICOMOON CSS */` and `/* END GENERATED ICOMOON CSS */` for example

foxbunny replied:
I think that might be a separate PR. If we take this further, we're also relying on icomoon as an external service to convert the icon. Ideally you'd just generate the icon font locally from a folder containing the icon files and do that as part of the webpack build.

foxbunny replied:
Basically, I suggest we just move with the codepen for now and optimize the workflow later.

foxbunny replied:
You mean like duplicate the existing declaration specifically for the placeholder?

foxbunny replied:
The reason I put it there is because `::before` selectors cannot be extended, btw.

foxbunny replied:
Pushed some changes. Let me know if that's what you meant.

ThiefMaster replied:
yes, exactly :)

ThiefMaster commented:
I pinged you in the missed threads

foxbunny commented:
Gotcha, thanks!

ThiefMaster commented on indico/modules/core/client/js/session_bar.js:
```diff
@@ -80,6 +71,7 @@
       },
       events: {
         show() {
+          // eslint-disable-next-line no-undef
```
You can avoid both `eslint-disable-next-line` statements by importing lodash properly ;) `import _ from 'lodash';`

foxbunny replied:
Hm. So why wasn't it imported to begin with? 🤔

foxbunny replied:
Anyway, I wouldn't dwell on this cause it's going away once the pull-down menu hits the PR backlog.

ThiefMaster replied:
bad old legacy times ;)

ThiefMaster commented on indico/web/client/styles/design_system/_form_controls.scss:
```diff
@@ -0,0 +1,69 @@
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+@use "utils";
+@use "layout";
```
```suggestion
@use 'utils';
@use 'layout';
```

ThiefMaster replied:
not sure why stylelint doen't catch that...

foxbunny replied:
Yeah, weird.

ThiefMaster replied:
probably related to that - maybe they already removed it from the standard ruleset: https://stylelint.io/migration-guide/to-15

foxbunny replied:
I'll solve the problem on my end.

<img width="239" alt="image" src="https://github.com/indico/indico/assets/65413/e1d0dbe8-d326-4cb7-8c17-05a5352e985e">

ThiefMaster commented on indico/web/client/styles/partials/_buttons.scss:
```diff
@@ -164,7 +164,7 @@ $i-button-spacing: 0.3rem;
   & ~ .i-button,
   & ~ .group {
-    margin-left: $i-button-spacing;
+    margin-right: $i-button-spacing;
```
Are you sure this change is correct and intentional?

Before:
![image](https://github.com/indico/indico/assets/179599/a22d14bb-4f88-4813-a35a-ad92b6c3d771)

After:
![image](https://github.com/indico/indico/assets/179599/26eafb4f-c623-4e64-ad4e-04dc501c5fa4)


So it's removing spacing that was meant to be there to separate buttons/groups

foxbunny replied:
Hmmmm. Probably unintentional.

foxbunny replied:
Reverted

ThiefMaster commented on indico/web/client/styles/partials/_buttons.scss:
```diff
@@ -289,7 +289,7 @@ a.i-big-button {
 
 .i-button.signin {
```
you can remove this whole definition since the button no longer exists. it was from the legacy login page (59afb5b37) which no longer exists
